### ISSUE TITLE
Compile error

### ISSUE THREAD

**Issue by Jose76Tron** [sigsegv-mvm/sigsegv-mvm#3]
I compiled the last commit and it worked, im really happy to see it working but i have a new problem, when i try to add the engiebot_dispensers.cpp i get a compile error

> 1>   Creating library C:\Users\amon\Desktop\Todo\sigsegv-mvm-master\MSVC\Binary\sigsegv.ext.2.tf2.lib and object C:\Users\amon\Desktop\Todo\sigsegv-mvm-master\MSVC\Binary\sigsegv.ext.2.tf2.exp
1>engiebot_dispensers.obj : error LNK2001: unresolved external symbol "private: static class MemberFuncThunk<class CTFNavMesh *,void,class CUtlVector<class CBaseObject *,class CUtlMemory<class CBaseObject *,int> > *,int> CTFNavMesh::ft_CollectBuiltObjects" (?ft_CollectBuiltObjects@CTFNavMesh@@0V?$MemberFuncThunk@PAVCTFNavMesh@@XPAV?$CUtlVector@PAVCBaseObject@@V?$CUtlMemory@PAVCBaseObject@@H@@@@H@@A)
1>engiebot_dispensers.obj : error LNK2001: unresolved external symbol "class GlobalThunk<class CTFNavMesh *> TheNavMesh" (?TheNavMesh@@3V?$GlobalThunk@PAVCTFNavMesh@@@@A)
1>engiebot_dispensers.obj : error LNK2001: unresolved external symbol "public: __thiscall PathFollower::PathFollower(void)" (??0PathFollower@@QAE@XZ)
1>engiebot_dispensers.obj : error LNK2001: unresolved external symbol "public: virtual __thiscall PathFollower::~PathFollower(void)" (??1PathFollower@@UAE@XZ)
1>engiebot_dispensers.obj : error LNK2001: unresolved external symbol "public: bool __thiscall Path::Compute<class CTFBotPathCost>(class INextBot *,class Vector const &,class CTFBotPathCost &,float,bool)" (??$Compute@VCTFBotPathCost@@@Path@@QAE_NPAVINextBot@@ABVVector@@AAVCTFBotPathCost@@M_N@Z)
1>C:\Users\amon\Desktop\Todo\sigsegv-mvm-master\MSVC\Binary\sigsegv.ext.2.tf2.dll : fatal error LNK1120: 5 unresolved externals

**Comment by sigsegv-mvm** [comment]
Well...

The proximate cause of this issue is that you need to also add `src/stub/nav.cpp` and `src/re/path.cpp` to the project for it to link properly. However, that will open up a bunch of additional issues, and it still won't build successfully.

----

Basically, the crux of the matter is that most of the sub-mods within this SM extension only really work with Linux srcds; and the ones that do work with Windows are generally already included in the MSVC project.

And that isn't due to any particular bias toward Linux on my part; it just happens to be far, far easier to get things hooked up to the Linux version of the dedicated server than the Windows version. That's almost entirely due to the fact that the Linux binaries (`server_srv.so` etc) contain symbol information (strings in the binary that make it super-easy to find the addresses of functions simply by looking up their name); whereas the Windows binaries (`server.dll` etc) do not have symbol information (meaning that for *every single* function and other code-related thing that I need to hook up to make the mods work, I have to find an alternative method of locating the relevant addresses, instead of simply looking them up by name).

Reliably getting the address of a function in the Linux dedicated server is as easy as finding the symbolic name of that function. It's fast, it only has to be done once, and it will almost never break in future game updates. So for example, if you open up `gamedata/sigsegv/client.txt` and look up the entry for `[client] C_BaseEntity::EstimateAbsVelocity`, the line `sym "_ZN12C_BaseEntity19EstimateAbsVelocityER6Vector"` sets up that association for the Linux server.

But to reliably get the address of a function on the Windows dedicated server, I typically have to come up with a pattern and mask of x86 instruction bytes from the function that uniquely matches that particular function and no other ones (see the `seek` and `mask` lines in that same gamedata entry from earlier for an example). Doing this in a robust way that has a maximum chance of not breaking in future game updates takes a ton of manual work compared to just copying the symbol string over; and by its very nature it is very likely to break in future game updates, because the compiler may make some small changes to the instructions in the function which doesn't change its actual functionality but which nevertheless completely break the pattern matching, which then means I have to come up with a new pattern and mask all over again.

And there are some other miscellaneous reasons why Windows support is harder than Linux support; I won't go into details here, but just as an example, the way the RTTI works is completely different with MSVC vs GCC.

----

So, to summarize, this means that setting up everything to work with Linux *and* Windows is a ***substantially*** greater amount of work than setting up everything to work with Linux and *just some selected things* to work with Windows. And as a consequence, there are many sub-mods in this extension which simply didn't seem worth the effort for me to get them to work on Windows.

Your options, then, are basically:
1. Do a ton of work to get the particular sub-mods you want working on Windows.
2. Build the extension on Linux, fire up a Linux srcds instance, and run it there instead.

Personally, I'd strongly recommend option 2. :neutral_face:

**Comment by Jose76Tron** [comment]
Thanks for answering.
So can i compile the extension and run the server on a virtual machine of linux in windows 7? and if so how i can compile the plugin on the linux version?

**Comment by marinesciencedude** [comment]
> Your options, then, are basically:
>
>   1.  Do a ton of work to get the particular sub-mods you want working on Windows.
>    2. Build the extension on Linux, fire up a Linux srcds instance, and run it there instead.
>
> Personally, I'd strongly recommend option 2. 😐

Out of interest, what exactly does one need to do for option 1? (otherwise, how can I ensure the Virtual Machine (VirtualBox) isn't slow, if this is a place I can ask help for it. Tried Ubuntu: near-impossible and Debian: not manageable. Also switched on Virtualisation for 64-bit Linux.)
